include/merlin/core_kernels/lookup_kernels.cuh

```diff
@@ -0,0 +1,722 @@
+  int groupID = threadIdx.x / GROUP_SIZE;
+  int rank = g.thread_rank();
+  int key_idx_base = (blockIdx.x * blockDim.x) + groupID * GROUP_SIZE;
+  if (key_idx_base >= n) return;
+  int loop_num =
+      (n - key_idx_base) < GROUP_SIZE ? (n - key_idx_base) : GROUP_SIZE;
+  if (rank < loop_num) {
+    int idx_block = groupID * GROUP_SIZE + rank;
+    K target_key = keys[key_idx_base + rank];
+    sm_target_keys[idx_block] = target_key;
+    K hashed_key = Murmur3HashDevice(target_key);
+    uint8_t target_digest = static_cast<uint8_t>(hashed_key >> 32);
+    sm_target_digests[idx_block] = static_cast<uint32_t>(target_digest);
+    int global_idx = hashed_key % (buckets_num * BUCKET_SIZE);
+    int bkt_idx = global_idx / BUCKET_SIZE;
+    const Bucket<K, V, S>* bucket = buckets + bkt_idx;
+    __pipeline_memcpy_async(sm_keys_ptr + idx_block, &(bucket->keys_),
+                            sizeof(K*));
+    __pipeline_commit();
+    __pipeline_memcpy_async(sm_values_ptr + idx_block, &(bucket->vectors),
+                            sizeof(V*));
+  }
+  __pipeline_wait_prior(0);
+
+  // Pipeline loading
+  S* scores_ptr =
+      reinterpret_cast<S*>(sm_keys_ptr[groupID * GROUP_SIZE] + BUCKET_SIZE);
+  uint8_t* digests_ptr = reinterpret_cast<uint8_t*>(scores_ptr + BUCKET_SIZE);
+  __pipeline_memcpy_async(sm_probing_digests[0] + groupID * DIGEST_SPAN + rank,
+                          digests_ptr + rank * 4, sizeof(uint32_t));
+  __pipeline_commit();
+  __pipeline_commit();  // padding
+  __pipeline_commit();  // padding
+
+  for (int i = 0; i < loop_num; i++) {
+    int key_idx_block = groupID * GROUP_SIZE + i;
+
+    /* Step1: prefetch all digests in one bucket */
+    if ((i + 1) < loop_num) {
+      S* scores_ptr =
+          reinterpret_cast<S*>(sm_keys_ptr[key_idx_block + 1] + BUCKET_SIZE);
+      uint8_t* digests_ptr =
+          reinterpret_cast<uint8_t*>(scores_ptr + BUCKET_SIZE);
+      __pipeline_memcpy_async(
+          sm_probing_digests[DIFF_BUF(i)] + groupID * DIGEST_SPAN + rank,
+          digests_ptr + rank * 4, sizeof(uint32_t));
+    }
+    __pipeline_commit();
+
+    /* Step2: check digests and load possible keys */
+    uint32_t target_digest = sm_target_digests[key_idx_block];
+    uint32_t target_digests = __byte_perm(target_digest, target_digest, 0x0000);
+    sm_counts[key_idx_block] = 0;
+    __pipeline_wait_prior(3);
+    uint32_t probing_digests =
+        sm_probing_digests[SAME_BUF(i)][groupID * DIGEST_SPAN + rank];
+    uint32_t find_result_ = __vcmpeq4(probing_digests, target_digests);
+    uint32_t find_result = 0;
+    if ((find_result_ & 0x01) != 0) find_result |= 0x01;
+    if ((find_result_ & 0x0100) != 0) find_result |= 0x02;
+    if ((find_result_ & 0x010000) != 0) find_result |= 0x04;
+    if ((find_result_ & 0x01000000) != 0) find_result |= 0x08;
+    int find_number = __popc(find_result);
+    int group_base = 0;
+    if (find_number > 0) {
+      group_base = atomicAdd(sm_counts + key_idx_block, find_number);
```

Comment:
It looks like the `atomicAdd_block` is enough here?

Reply:
Yes, I agree with you. `__atomicAdd_block` is more proper.

Reply:
However, when I use `__atomicAdd_block`， I got the error:`identifier "__atomicAdd_block" is undefined`.
I think its related to CMakeLists.txt : `set_target_properties(merlin_memory_pool_test PROPERTIES  CUDA_ARCHITECTURES OFF)`.
And, accorrding to the CUDA Doc, atomicAdd support shared memory, so I think use atomicAdd is the cheapest way at present.

Reply:
atomicAdd works for both global and shared memory